File: Python/Medium/2438_Range_Product_Queries_of_Powers/solution.py
class Solution:
    def productQueries(self, n: int, queries: list[list[int]]) -> list[int]:
        bNum = bin(n)
        powers = []

        bCount = 0
        for i in range(len(bNum)-1, 0, -1):
            if bNum[i] == "b":
                break
            if bNum[i] == "0":
                bCount += 1
                continue
            powers.append(2**bCount)
            bCount += 1

        output = []
        mod = 10**9 + 7
        for q in queries:
            start, stop = q
            count = 1
            for i in range(start, stop + 1):
                count *= powers[i]
            count = count % mod
            output.append(count)
        return output

File: Golang/Medium/2438_Range_Product_Queries_of_Powers/solution.go
package rangeproductqueriesofpowers

import (
	"math"
	"strconv"
)

func productQueries(n int, queries [][]int) []int {
	bNum := strconv.FormatInt(int64(n), 2)

	powers := []int{}
	count := 0
	for i := len(bNum) - 1; i >= 0; i-- {
		num := string(bNum[i])
		if num == "1" {
			v := math.Pow(2.0, float64(count))
			powers = append(powers, int(v))
		}
		count += 1
	}

	const mod = 1000000007
	output := []int{}
	for _, q := range queries {
		v := 1
		start := q[0]
		stop := q[1]
		for i := start; i <= stop; i++ {
			v = (v * powers[i]) % mod
		}
		output = append(output, v)
	}
	return output
}
